### PR TITLE
Correctly format files with dos line endings

### LINKED ISSUE #80: Formatter changing file format from dos -> unix
I am running windows and found `leptosfmt` is changing file format from dos(CRLF) -> unix(LF). 
Snippet below. Ran command `leptosfmt example.rs`
```rust
use leptos::*;

#[component]
fn App() -> impl IntoView {
    view! { 
        <h1>"Iteration"</h1> 
        <h1>"UnIteration"</h1> 
    }
}


fn main() {
    leptos::mount_to_body(|| view! { <App/> })
}
```
Before:
image

After:
image

One interesting thing, if the `view!` macro had only one component the file format is not being changed. Snippet below
```rust
use leptos::*;

#[component]
fn App() -> impl IntoView {
    view! { 
        <h1>"Iteration"</h1> 
    }
}


fn main() {
    leptos::mount_to_body(|| view! { <App/> })
}

```

### PR DESCRIPTION
**Pull request by nikhilraojl**
Closes #80 

I did this the simplest way, but it may not be the right way. Please suggest if something can be improved or changed

### CODE REVIEW

File: printer/src/algorithm.rs [hunk 1]
```diff
@@ -78,6 +78,8 @@ pub struct Printer {
     indent: usize,
     // Buffered indentation to avoid writing trailing whitespace
     pending_indentation: usize,
+    // To correctly print with dos line endings(CRLF)
```
Comment (bram209) [inline printer/src/algorithm.rs:81]:
```suggestion
    // Print CRLF line ending instead of LF
```

File: printer/src/algorithm.rs [hunk 2]
```diff
@@ -99,9 +101,14 @@ impl Printer {
             print_stack: Vec::new(),
             indent: 0,
             pending_indentation: 0,
+            dos_line_endings: false,
```
Comment (bram209) [inline printer/src/algorithm.rs:104]:
```suggestion
            dos_line_endings: settings.dos_line_endings,
```

File: printer/src/algorithm.rs [hunk 3]
```diff
@@ -78,6 +78,8 @@ pub struct Printer {
     indent: usize,
     // Buffered indentation to avoid writing trailing whitespace
     pending_indentation: usize,
+    // To correctly print with dos line endings(CRLF)
+    dos_line_endings: bool,
```
Comment (bram209) [inline printer/src/algorithm.rs:82]: How about making `CRLF` more explicit by renaming to:
```suggestion
    crlf_line_endings: bool,
```

File: printer/src/algorithm.rs [hunk 4]
```diff
@@ -99,9 +101,14 @@ impl Printer {
             print_stack: Vec::new(),
             indent: 0,
             pending_indentation: 0,
+            dos_line_endings: false,
         }
     }
 
+    pub fn set_dos(&mut self, value: bool) {
```
Comment (bram209) [inline printer/src/algorithm.rs:108]: we could pass this to the instructor of `Printer` instead, no need for a setter in that case.
(I don't think you would ever switch to a different line ending style halfway formatting)

File: formatter/src/formatter/mac.rs [hunk 5]
```diff
@@ -162,6 +162,13 @@ pub fn format_macro(
                 source,
                 mac.mac.tokens.clone(),
             );
+            let dos_endings = source
```
Comment (bram209) [inline formatter/src/formatter/mac.rs:165]:
```suggestion
            let crlf_line_endings = source
```

File: formatter/src/formatter/mac.rs [hunk 6]
```diff
@@ -162,6 +162,13 @@ pub fn format_macro(
...
+                .raw_lines()
+                .nth(0)
+                .unwrap()
+                .to_string()
+                .ends_with("\r\n");
+            printer.set_dos(dos_endings);
```
Comment (bram209) [inline formatter/src/formatter/mac.rs:171]:
```suggestion
             let mut printer = Printer::new(PrinterSettings { ..settings.into(), crlf_line_endings });
```

File: formatter/src/formatter/mac.rs [hunk 7]
```diff
@@ -162,6 +162,13 @@ pub fn format_macro(
...
             Formatter::with_source(*settings, &mut printer, source, whitespace)
         }
         None => Formatter::new(*settings, &mut printer),
```
Comment (bram209) [inline formatter/src/formatter/mac.rs:174]:
```suggestion
        None => {
            let mut printer = Printer::new(settings.into());
            Formatter::new(*settings, &mut printer)
        },
```

**Comment by bram209** [comment]
hey, thanks for your contribution!

I am not really a fan of platform specific line endings. Files can be easily transferred from one platform to the other, etc. And I believe on the reading side every editor would be able to read LF line endings? I haven't use windows in ages, but I think you could configure git to automatically convert line endings to LF.

Anyhow, it is still nice to support this. Code looks good to me, only have some very minor suggestions (mostly naming), well done 👍

**Comment by bram209** [comment]
I think most people will be fine with having the formatter automatically detect line endings, but if you feel like it, you could also add a configuration option `newline_style` to `FormatterSettings` like so:

https://github.com/bram209/leptosfmt/blob/main/formatter/src/formatter/mod.rs#L38

This could have the following variants: "Auto", "Native", "Windows", "Unix".

Where "Native" you could use conditional compilation, like so:
```rust
val crlf_line_endings =  cfg!(windows) 
```

But if you want to leave it at what you currently have, that is perfectly fine as well : )

**Comment by nikhilraojl** [comment]
> I haven't use windows in ages, but I think you could configure git to automatically convert line endings to LF.

CRLF is default on windows. But git will replace CRLF as LF when a commit is done, but files will stay as CRLF for editing locally. I understand this is not good but that's windows 🤷 

> I think most people will be fine with having the formatter automatically detect line endings, but if you feel like it, you could also add a configuration option newline_style to FormatterSettings like so:

I would like to do this later as I don't conditional compilation stuff yet. Will definitely give it a go alter

File: formatter/src/formatter/mac.rs (hunk 6 above)
Comment (nikhilraojl) [reply formatter/src/formatter/mac.rs]: `crlf_line_endings` moved from `Printer` to `PrinterSettings` struct as per suggestion

File: formatter/src/formatter/mac.rs [hunk 8]
```diff
@@ -155,16 +155,29 @@ pub fn format_macro(
...
             let whitespace = crate::collect_comments::extract_whitespace_and_comments(
                 source,
                 mac.mac.tokens.clone(),
             );
+            let crlf_line_endings = source
+                .raw_lines()
+                .nth(0)
+                .unwrap()
+                .to_string()
+                .ends_with("\r\n");
```
Comment (bram209) [inline formatter/src/formatter/mac.rs:170]: Let's avoid panicking if the file is empty
```suggestion
            let crlf_line_endings = source
                .raw_lines()
                .nth(0)
                .map(|raw_line| raw_line.to_string().ends_with("\r\n"))
                .unwrap_or_default();
```

Comment (nikhilraojl) [reply formatter/src/formatter/mac.rs]: Done

**Review by bram209 — APPROVED** [review]

**Pull request merged**